quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,69 +196,173 @@ impl Recv {
+                        self.pending.post_read(
+                            self.id,
+                            ShouldTransmit(false),
+                            ShouldTransmit(false),
+                            max_dirty,
+                        );
+                        self.fused = Some(ReadStatus::Finished);
+                        ReadStatus::Finished
+                    } else {
+                        self.streams.recv.insert(self.id, self.recv.take().unwrap());
+                        self.fused = Some(ReadStatus::Blocked);
+                        ReadStatus::Blocked
+                    }
+                }
+                RecvState::ResetRecvd { error_code, .. } => {
+                    rs.state = RecvState::Closed;
+                    self.streams.stream_freed(self.id, StreamHalf::Recv);
+                    let max_dirty = self.streams.take_max_streams_dirty(self.id.dir());
+                    self.pending.post_read(
+                        self.id,
+                        ShouldTransmit(false),
+                        ShouldTransmit(false),
+                        max_dirty,
+                    );
+                    self.fused = Some(ReadStatus::Reset(error_code));
+                    ReadStatus::Reset(error_code)
+                }
+                RecvState::Closed => {
+                    self.fused = Some(ReadStatus::Closed);
+                    ReadStatus::Closed
+                }
+            },
+        };
+
+        ReadResult {
+            chunk,
+            status,
+            should_transmit,
+        }
+    }
+}
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
+/// Errors that might occur when opening up a reader for a stream
+#[derive(Clone, Copy, Debug, Error, PartialEq, Eq, Hash)]
+pub enum ReadableError {
+    /// Attempted an ordered read following an unordered read
+    ///
+    /// Performing an unordered read allows discontinuities to arise in the receive buffer of a
+    /// stream which cannot be recovered, making further ordered reads impossible.
+    #[error("ordered read after unordered read")]
+    IllegalOrderedRead,
+    /// The stream has not been opened or was already stopped, finished, or reset
+    #[error("unknown stream")]
+    UnknownStream,
 }
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+impl From<IllegalOrderedRead> for ReadableError {
+    fn from(_: IllegalOrderedRead) -> Self {
+        ReadableError::IllegalOrderedRead
     }
 }
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+/// Results of reading from a stream
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub struct ReadResult {
+    /// The data read from the stream
+    pub chunk: Option<Chunk>,
+    /// The status of the stream after reading
+    pub status: ReadStatus,
+    /// Whether the connection is ready to transmit
+    pub should_transmit: ShouldTransmit,
```

Comment:
Could we yield this once at the end, rather than per call to `next`? I understand that you want to make it hard to skip, but there's no way to enforce that a struct field gets looked at, and it seems strange to check this repeatedly when we can only act on it after we're done due to the borrow.